ifsbench/benchmark.py

```diff
@@ -0,0 +1,177 @@
+# (C) Copyright 2020- ECMWF.
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+from dataclasses import dataclass, field
+import os
+from pathlib import Path
+from typing import List, Optional
+
+from ifsbench.application import Application
+from ifsbench.arch import Arch
+from ifsbench.data import DataHandler
+from ifsbench.env import EnvHandler, DefaultEnvPipeline
+from ifsbench.job import Job
+from ifsbench.launch import Launcher
+
+
+__all__ = ['DefaultBenchmarkSetup', 'DefaultBenchmark']
+
+@dataclass
+class DefaultBenchmarkSetup:
+    """
+    Generic benchmark setup.
+
+    This dataclass holds generic information that is needed by benchmarks.
+    """
+
+    #: The data pipeline for the initial data setup.
+    data_handlers:  List[DataHandler] = field(default_factory=list)
+
+    #: Environment handlers that are used for the initial data setup.
+    env_handlers:  List[EnvHandler] = field(default_factory=list)
+
+    #: The application that gets benchmarked.
+    application: Optional[Application] = None
+
+    def merge(self, other):
+        """
+        Create a copy of this object and merge it with another given setup
+        object by
+        * appending the data_handlers
+        * appending the env_handlers
+        * replacing the application with ``other.application`` (if specified).
+
+        Parameters
+        ----------
+        other: DefaultBenchmarkSetup
+            The other setup object that is merged into this one.
+
+        Returns
+        -------
+        The merged object.
+        """
+        data_handlers = self.data_handlers + other.data_handlers
+        env_handlers = self.env_handlers + other.env_handlers
+        application = self.application
+        if other.application:
+            application = other.application
```

Comment:
why does 'other' take precedence if both are provided? I think in a member function I would expect that the other gets merged into self, so actually maybe this is better as a function outside the class as `def merge(first, second) -> DefaultBenchmarkSetup` to make it clear that it returns a new object rather than merging into the current.